sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,206 @@
+# Author: Henry Lin <hlin117@gmail.com>
+
+from __future__ import division, absolute_import
+
+import numbers
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_array, check_is_fitted, column_or_1d
+
+
+class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+    """Bins continuous data into k equal width intervals.
+
+    Parameters
+    ----------
+    n_bins : int or array-like, shape (n_features_,) (default=2)
+        The number of bins to produce. The intervals for the bins are
+        determined by the minimum and maximum of the input data.
+        Raises ValueError if n_bins < 2.
+
+        If n_bins is an array, and there is a categorical feature at index i,
+        n_bins[i] will be ignored.
+
+    categorical_features : int array-like (default=None)
+        Column indices of categorical features. Categorical features
+        will not be discretized and will be ignored by the transformation.
+
+    Attributes
+    ----------
+    mins_ : float array, shape (n_features_,)
+        Minimum value per feature in X.
+
+    maxes_ : float array, shape (n_features_,)
+        Maximum value per feature in X.
+
+    n_features_ : int
+        Number of features in X.
+
+    Example
+    -------
+    >>> X = [[-2, 1, -4,   -1], \
+             [-1, 2, -3, -0.5], \
+             [ 0, 3, -2,  0.5], \
+             [ 1, 4, -1,    2]]
+    >>> dis = KBinsDiscretizer(n_bins=3)
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+    >>> dis.transform(X) # doctest: +SKIP
+    array([[0, 0, 0, 0],
+           [1, 1, 1, 0],
+           [2, 2, 2, 1],
+           [2, 2, 2, 2]])
+
+    Categorical features can be specified in the constructor.
+
+    >>> dis = KBinsDiscretizer(n_bins=3, categorical_features=[1])
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+
+    Discretization will ignore categorical features.
+    >>> dis.transform(X) # doctest: +SKIP
+    array([[ 0., 1., 0., 0.],
+           [ 1., 2., 1., 0.],
+           [ 2., 3., 2., 1.],
+           [ 2., 4., 2., 2.]])
+    """
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+    def fit(self, X, y=None):
+        """Fits the estimator.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            Data to be discretized.
+
+        Returns
+        -------
+        self
+        """
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Reshape your data.")
+
+        n_features = X.shape[1]
+        self._check_categorical_features(self.categorical_features, X.shape[1])
+        self._check_n_bins(self.n_bins, n_features)
+
+        self.n_features_ = n_features
+        self.mins_ = np.min(X, axis=0)
```

Comment:
Since I manipulate the `maxes_` attribute on line 182, I thought it was best to keep `maxes_`. Also, I'm not actually sure what `ptp` stands for. (Though I know you want `ptp_ = max_ - min_`.

I'll do as you say, and rename `mins_` to `min_` and `maxes_` to `max_`.